pkg/registry/core/persistentvolume/strategy.go

```diff
@@ -134,11 +137,22 @@ func (persistentvolumeStatusStrategy) GetResetFields() map[fieldpath.APIVersion]
 	return fields
 }
 
+var nowFunc = metav1.Now
+
 // PrepareForUpdate sets the Spec field which is not allowed to be changed when updating a PV's Status
 func (persistentvolumeStatusStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object) {
+	now := nowFunc()
 	newPv := obj.(*api.PersistentVolume)
 	oldPv := old.(*api.PersistentVolume)
 	newPv.Spec = oldPv.Spec
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.PersistentVolumeLastPhaseTransitionTime) {
+		if oldPv.Status.Phase != newPv.Status.Phase {
```

Comment:
It's surprising to stomp a client-set value unconditionally.

I expected the server to only populate the field if the client was 1) clearing it on update or 2) not changing it when the phase changed. 

Something like this:

```go
switch {
case oldPv.Status.Phase == newPv.Status.Phase && newPv.Status.LastPhaseTransitionTime == nil:
  // phase didn't change, preserve the existing transition time if set
  newPv.Status.LastPhaseTransitionTime = oldPv.Status.LastPhaseTransitionTime

case oldPv.Status.Phase != newPv.Status.Phase && (newPv.Status.LastPhaseTransitionTime == nil || newPv.Status.LastPhaseTransitionTime.Equal(oldPv.Status.LastPhaseTransitionTime)):
  // phase changed and client didn't set or didn't change the transition time
  now := nowFunc()
  newPv.Status.LastPhaseTransitionTime = &now
}
```

Reply:
is the field expected to be populated on create as well if empty?

Reply:
We had long discussion with @msau42 who / when should set the field and we came up with the current code.
Clients can set it to whatever they want, but the API server overwrites it if and only if the phase changes.

Reply:
one of good arguments was that the timestamp will be close to audit events

Reply:
why is this field special in the server overwriting the client-set time, when compared to other things like pod timestamps set by the kubelet, condition timestamps set by the controller, etc?

Reply:
It's not that special. So shall we move changing the field back to PV controller, i.e. slip 1.28?

Reply:
I think the suggestion in https://github.com/kubernetes/kubernetes/pull/116469#discussion_r1270744169 would still populate and update the field for a controller that did not set the timestamp

Reply:
@liggitt I've updated the PR to reflect this behavior. Only the 3rd test case changed so that the client provided value is preserved, the rest is the same - I suppose that's the expected behavior now?

Reply:
There's one more question on if the timestamp should be initialized on create as well. I think it would be nice to have so that any tooling that is looking at this only needs to look at one field instead of special casing the creation case.

Reply:
Create appears to be forcing an empty status and phase, so leaving the time stamp empty seems ok?

Reply:
Remind me again when does defaulting happen in relation to `PrepareForCreate()` and `Validate()`. We have logic to default the phase to "Pending": https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/v1/defaults.go#L300

Reply:
Huh... defaulting happens on the incoming object before PrepareForCreate, so clearing status will wipe out the default

When the newly created object is next read from etcd, defaulting will fill the status.phase field again. If you want the newly created object to have phase="Pending" and lastTransitionTimestamp=now(), that seems ok to set in PrepareForCreate()